arangod/VocBase/Methods/AqlUserFunctions.cpp

```diff
@@ -0,0 +1,445 @@
+                  name +
+                  "' is not a valid name.");
+  }
+
+  std::string code;
+  std::string tmp;
+  try {
+    auto cvString = userFunction.get("code");
+    if (!cvString.isString()) {
+      return Result(TRI_ERROR_QUERY_FUNCTION_INVALID_CODE,
+		    "expecting function or string");
+    }
+    tmp = cvString.copyString();
+
+    if (tmp.empty()) {
+      return Result(TRI_ERROR_QUERY_FUNCTION_INVALID_CODE,
+		    "expecting function - must not be empty.");
+    }
+    code = std::string("(") + tmp + "\n)";
+  }
+  catch (...) {
+      return Result(TRI_ERROR_QUERY_FUNCTION_INVALID_CODE,
+		    "missing mandatory property - 'code'");
+  }
+  
+  bool isDeterministic = false;
+  try {
+    isDeterministic = userFunction.get("isDeterministic").getBool();
+  }
+  catch (...) {}
+
+  {
+    ISOLATE;
+    bool throwV8Exception = (isolate != nullptr);
+    V8ContextDealerGuard dealerGuard(res, isolate, vocbase, true /*allowModification*/);
+    if(res.fail()){
+      return res;
+    }
+
+    std::string testCode("(function() { var callback = ");
+    testCode += tmp + "; return callback; })()";
+    v8::HandleScope scope(isolate);
+
+
+    v8::Handle<v8::Value> result;
+    {
+      v8::TryCatch tryCatch;
+
+      result = TRI_ExecuteJavaScriptString(isolate,
+                                           isolate->GetCurrentContext(),
+                                           TRI_V8_STD_STRING(isolate, testCode),
+                                           TRI_V8_ASCII_STRING(isolate, "userFunction"),
+                                           false);
+
+      int state = 0;
+      if (result.IsEmpty()) {
+        state = 1;
+      }
+      else if (tryCatch.HasCaught()) {
+        state = 2;
+      }
+      else if (!result->IsFunction()) {
+        state = 3;
+      }
+
+      if (state != 0) {
+        const std::string msg[] = {
+          "",
+          "Empty result.",
+          "Exception occured",
+          "Is not a function"
+        };
+        res.reset(TRI_ERROR_QUERY_FUNCTION_INVALID_CODE,
+                  TRI_StringifyV8Exception(isolate, &tryCatch) +
+                  msg[state]);
+        if (!tryCatch.CanContinue()) {
+          if (throwV8Exception) {
+            tryCatch.ReThrow();
+          }
+          TRI_GET_GLOBALS();
+          v8g->_canceled = true;
+        }
+      }
+    }
+  }
+  if (!res.ok()) {
+    return res;
+  }
+  std::string _key(name);
+  basics::StringUtils::toupperInPlace(&_key);
+
+  VPackBuilder oneFunctionDocument;
+  oneFunctionDocument.openObject();
+  oneFunctionDocument.add("_key", VPackValue(_key));
+  oneFunctionDocument.add("name", VPackValue(name));
+  oneFunctionDocument.add("code", VPackValue(code));
+  oneFunctionDocument.add("isDeterministic", VPackValue(isDeterministic));
+  oneFunctionDocument.close();
+
+  arangodb::OperationOptions opOptions;
+  opOptions.isRestore = false;
+  opOptions.waitForSync = true;
+  opOptions.returnNew = false;
```

Comment:
returnNew = false should be the default behavior and does not need to be set explicitly.